radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/transactions/TransactionStatusService.java

```diff
@@ -62,28 +62,64 @@
  * permissions under this License.
  */
 
-package com.radixdlt.store;
+package com.radixdlt.api.transactions;
 
-import com.radixdlt.atom.Txn;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.inject.Inject;
+import com.radixdlt.api.data.TransactionStatus;
+import com.radixdlt.api.store.berkeley.BerkeleyTransactionsByIdStore;
+import com.radixdlt.environment.EventProcessor;
 import com.radixdlt.identifiers.AID;
+import com.radixdlt.mempool.MempoolAddFailure;
+import com.radixdlt.mempool.MempoolAddSuccess;
+import org.json.JSONObject;
 
+import java.time.Duration;
 import java.util.Optional;
 
-/**
- * Index of atoms
- */
-public interface TxnIndex {
-	/**
-	 * Checks whether the given aid is contained in this view
-	 * @param aid The aid
-	 * @return Whether the given aid is contained in this view
-	 */
-	boolean contains(AID aid);
+import static com.radixdlt.api.data.TransactionStatus.CONFIRMED;
+import static com.radixdlt.api.data.TransactionStatus.FAILED;
+import static com.radixdlt.api.data.TransactionStatus.PENDING;
+import static com.radixdlt.api.data.TransactionStatus.TRANSACTION_NOT_FOUND;
+
+public class TransactionStatusService {
+	private final Cache<AID, TransactionStatus> cache = CacheBuilder.newBuilder()
+		.maximumSize(100000)
+		.expireAfterAccess(Duration.ofMinutes(10))
+		.build();
+	private final BerkeleyTransactionsByIdStore store;
+
+	@Inject
+	public TransactionStatusService(BerkeleyTransactionsByIdStore store) {
+		this.store = store;
+	}
+
+	private void onReject(MempoolAddFailure mempoolAddFailure) {
+		cache.put(mempoolAddFailure.getTxn().getId(), FAILED);
+	}
+
+	private void onSuccess(MempoolAddSuccess mempoolAddSuccess) {
+		cache.put(mempoolAddSuccess.getTxn().getId(), PENDING);
+	}
+
+	public EventProcessor<MempoolAddFailure> mempoolAddFailureEventProcessor() {
+		return this::onReject;
+	}
+
+	public EventProcessor<MempoolAddSuccess> mempoolAddSuccessEventProcessor() {
+		return this::onSuccess;
+	}
+
+	public Optional<JSONObject> getTransaction(AID txId) {
+		return store.getTransactionJSON(txId);
+	}
 
-	/**
-	 * Gets the atom associated with a certain aid
-	 * @param aid The aid
-	 * @return The atom associated with the given aid (if any)
-	 */
-	Optional<Txn> get(AID aid);
+	public TransactionStatus getTransactionStatus(AID txId) {
+		var status = cache.getIfPresent(txId);
+		if (status != CONFIRMED && store.contains(txId)) {
```

Comment:
Once stored in `BerkeleyTransactionsByIdStore`, does a transaction ever change its state? I remember you once mentioned we have have some re-branching logic.

Reply:
Nope, no re-branching logic at the moment.